### PR TITLE
MultiWi 2.3 PID controller port

### PR DESCRIPTION
**Pull request by MJ666**
MultiWii 2.3 PID controller 
set pid_controller = 3 (MultiWii 2.3 PID's)

MultiWii hybrid PID controller. Roll and pitch is using the MultiWii 2.2 algorithm and yaw is using the 2.3 algorithm. 
set pid_controller = 4 (MultiWii Hybrid PID's)

### CODE REVIEW

File: src/main/flight/flight.c [hunk 1]
```diff
@@ -287,8 +288,211 @@ static void pidMultiWii(pidProfile_t *pidProfile, controlRateConfig_t *controlRa
...
+    if(motorCount > 3) {
+        int16_t limit = GYRO_P_MAX - pidProfile->D8[FD_YAW];
+        PTerm = constrain(PTerm, -limit, +limit);
+    }
+
+    ITerm = constrain((int16_t)(errorGyroI[FD_YAW] >> 13), -GYRO_I_MAX, +GYRO_I_MAX);
+
+    axisPID[FD_YAW] =  PTerm + ITerm;
+
+#ifdef BLACKBOX
```
Comment (ledvinap) [inline src/main/flight/flight.c]: This code is misplaced - only YAW is stored

Comment (MJ666) [inline src/main/flight/flight.c]: Roll and pitch are stored in the loop. Line 358.

File: src/main/flight/flight.c [hunk 2]
```diff
@@ -287,8 +288,211 @@ static void pidMultiWii(pidProfile_t *pidProfile, controlRateConfig_t *controlRa
...
+
+            ITermACC = ((int32_t)errorAngleI[axis] * pidProfile->I8[PIDLEVEL]) >> 12;  // 32 bits is needed for calculation:10000*I8 could exceed 32768   16 bits is ok for result
+
+            ITerm = ITermACC + ((ITerm - ITermACC) * prop >> 9);
+            PTerm = PTermACC + ((PTerm - PTermACC) * prop >> 9);
+        }
+
+        PTerm -= ((int32_t)(gyroData[axis] / 4) * dynP8[axis]) >> 6;   // 32 bits is needed for calculation
+
+        delta = (gyroData[axis] - lastGyro[axis]) / 4;   // 16 bits is ok here, the dif between 2 consecutive gyro reads is limited to 800
```
Comment (ledvinap) [inline src/main/flight/flight.c]: Not important, but you can divide DTerm, gaining some accuracy

File: src/main/flight/flight.c (hunk 1 above)
Comment (ledvinap) [inline src/main/flight/flight.c]: But blackbox uses `axisPID_P[], axisPID_I[], axisPID_D[]`

**Comment by ledvinap** [comment]
Is there any advantage to hybrid version?

File: src/main/flight/flight.c (hunk 1 above)
Comment (MJ666) [inline src/main/flight/flight.c]: Committed the fix.

**Comment by hydra** [comment]
Ok, gonna merge this and build a pre-release tonight for people to test.  We can make any further adjustments to the master branch, i think is is ok since it's isolated from other code and users have to manually choose to use them.

**Comment by hydra** [comment]
@MJ666 can you create a new PR that updates the documentation please?

**Comment by MJ666** [comment]
I will do.

**Comment by hydra** [comment]
I flew this today, the yaw authority is like night and day between pid controller 0 and 3. I like it.

**Comment by MJ666** [comment]
Yaw was always our main pain point. For the MultiWii2.3 we use some additional tuning for the Nanos(Yawrate to 1.0). You should also try Harakiri.this will fly well already with mainly defaults (Yawrate to 0.15 - 0.2).

**Pull request merged**